hunter_base/include/hunter_base/hunter_messenger.hpp

```diff
@@ -211,18 +249,94 @@ class HunterMessenger {
     // ROS_INFO("Cmd received:%f, %f", msg->linear.x, msg->angular.z);
   }
 
-  // template <typename T,std::enable_if_t<!std::is_base_of<HunterRobot, T>::value,bool> = true>
-  void SetHunterMotionCommand(const geometry_msgs::msg::Twist::SharedPtr &msg) {
+  void odometryCallback(const nav_msgs::msg::Odometry::SharedPtr msg) {
+        v_ = msg->twist.twist.linear.x;
+        omega_ = msg->twist.twist.angular.z;
+  }
 
-    std::shared_ptr<HunterRobot> base;
+  void onParameterEvent(const rcl_interfaces::msg::ParameterEvent::SharedPtr event) {
+      for (auto & changed_parameter : event->changed_parameters) {
+          const std::string & name = changed_parameter.name;
+          if (name == "kp_v") {
+              kp_v_ = changed_parameter.value.double_value;
+              RCLCPP_INFO(node_->get_logger(), "Updated kp_v to: %f", kp_v_);
+          } else if (name == "kd_v") {
+              kd_v_ = changed_parameter.value.double_value;
+              RCLCPP_INFO(node_->get_logger(), "Updated kd_v to: %f", kd_v_);
+          } else if (name == "kp_w") {
+              kp_w_ = changed_parameter.value.double_value;
+              RCLCPP_INFO(node_->get_logger(), "Updated kp_w to: %f", kp_w_);
+          } else if (name == "kd_w") {
+              kd_w_ = changed_parameter.value.double_value;
+              RCLCPP_INFO(node_->get_logger(), "Updated kd_w to: %f", kd_w_);
+          }
+      }
+  }
+
+
+  // define the control law
+  void controlLoop(double& len_vel, double& anu_vel) {
+
+    auto current_time = node_->get_clock()->now();
+    double dt = 1.0/sim_control_rate_;
+    // double dt = (current_time - last_time_control_law_).seconds();
```

Comment:
@GPrathap for the PD controller dt should be calculated based on the call back of this function, so I think the commented-out line is the correct one! @LeonardoGuevara  what do you think?